compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs

```diff
@@ -1114,29 +1221,47 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
                     .iter()
                     .find(|vd| tcx.hygienic_eq(assoc_ident, vd.ident(tcx), adt_def.did()));
                 if let Some(variant_def) = variant_def {
-                    if permit_variants {
+                    if mode.permit_variants() {
                         tcx.check_stability(variant_def.def_id, Some(hir_ref_id), span, None);
                         let _ = self.prohibit_generic_args(
                             slice::from_ref(assoc_segment).iter(),
                             GenericsArgsErrExtend::EnumVariant { qself, assoc_segment, adt_def },
                         );
-                        return Ok((qself_ty, DefKind::Variant, variant_def.def_id));
+                        return Ok(LoweredAssoc::Variant {
+                            adt: qself_ty,
+                            variant_did: variant_def.def_id,
+                        });
                     } else {
                         variant_resolution = Some(variant_def.def_id);
                     }
                 }
             }
 
-            // FIXME(inherent_associated_types, #106719): Support self types other than ADTs.
-            if let Some((ty, did)) = self.probe_inherent_assoc_ty(
-                assoc_ident,
-                assoc_segment,
-                adt_def.did(),
-                qself_ty,
-                hir_ref_id,
-                span,
-            )? {
-                return Ok((ty, DefKind::AssocTy, did));
+            match mode {
```

Comment:
i.e. this is a pretty big match that could just be `probe_inherent_assoc_shared` and then returning the defid + args pair